Add support for Rust lang

ian-h-chamberlain opened the pull request:
Closes #132 

Here is my first take on adding support for Rust. It seems to work locally with `make dev` although I have run into some issues where the holes don't display on the main menu and I'm not sure how to test against a DB (or if that's really even necessary).

## Summary
* A few `bash` changes to try and make build scripts, etc. more portable. Everything still works for me after these changes on CentOS Linux and I can actually run the server on macOS as well. I wasn't able to easily test on Arch or other Linux distros,  however.
* Add Dockerfile for Rust lang. This is almost literally just the upstream `rust:1.40.0-alpine3.10` image – if we need to remove some of the extraneous stuff there I will need to dig in a little deeper to see what can/can't be excluded from the image, although in practice it looks like the image is around 330MB – smaller than some of the other language images but still on the larger side for sure:

| Image                | Size   |
|----------------------|--------|
| code-golf-bash       | 1.14MB |
| code-golf-brainfuck  | 1.56MB |
| code-golf-c          | 1.54MB |
| code-golf-haskell    | 946MB  |
| code-golf-j          | 3.45MB |
| code-golf-javascript | 27.3MB |
| code-golf-julia      | 282MB  |
| code-golf-lisp       | 35.2MB |
| code-golf-lua        | 273kB  |
| code-golf-nim        | 2.6GB  |
| code-golf-perl       | 3.23MB |
| code-golf-perl6      | 36.4MB |
| code-golf-php        | 10.5MB |
| code-golf-python     | 187MB  |
| code-golf-ruby       | 22.8MB |
| code-golf-rust       | 363MB  |

* Plumbing through the various pages + views. Hopefully I didn't miss anything here
* Add CodeMirror support for Rust. It's not as full featured as some of the others but seems to work okay in practice. Also added an SVG for the logo here – but I couldn't figure out if those are used anywhere??
* Mount `/dev` into the language's environment. This change I was less sure of, since I'm not sure I fully grasp what `run-lang` does, but it seems to work without breaking other languages. The Rust compiler requires `/dev/urandom` to be available, but it is not present in the rootfs, so this change mounts the existing `/dev` to the one in the rootfs. If you know of a different/more correct way to do this, let me know and I will gladly change it.

JRaspass commented:
This looks really good!

So the SVGs used to be used in the old UI, and I intend to make sure of them again soon, so thanks for adding that.

I'm not so keen about mounting the entire /dev into the container (unless I misread your ASM), could we just perhaps create /dev/urandom? `mknod` appears to be the way to go based on https://www.tldp.org/LDP/lfs/LFS-BOOK-6.1.1-HTML/chapter06/devices.html so I guess we want the syscall equiv of the cli tool - http://man7.org/linux/man-pages/man2/mknod.2.html . Let me know if that doesn't work but it feels a bit safer in terms of exposing less to the containers, I presume rust can't be compiled to use the getrandom syscall instead?

ian-h-chamberlain commented:
Okay, I've made a change to use `devtmpfs` in the same manner as `tmpfs` is being used for `/tmp` – this seems to work fine and I imagine is a bit safer so hopefully addresses your concerns.

Is there anything else you'd like me to address to make this mergeable? The only other thing I wasn't super sure of was the `compile_and_run.sh` wrapper – is it okay to just create a tmp executable and `exec` it, or is there a concern about leaving stuff lying around afterwards?

Also, this is more of a preference on the user experience, but should be leave backtraces enabled by default when running the code?

Thanks for the quick feedback, this ended up being pretty manageable to implement.

JRaspass commented:
Leaving stuff behind is fine, its a new file system namespace that will disappear afterwards.

I'm not 100% sold on devtmpfs, it is definitely better than the bind mount but I still worry about how much that exposes inside the container. I might just be being paranoid though. Did mknod not work, out of curiosity?

Other than that I've caused you a minor conflict with the `s/Perl 6/Raku/` rename so if you wouldn't mind rebasing.

JRaspass commented:
Sorry I missed the backtraces question, that's probably one for the rust users, I guess more debugging doesn't hurt.

ian-h-chamberlain commented:
Okay, I looked a bit more into this `urandom` issue, and found something kind of interesting.

`rustc` actually _does_ use `getrandom` by default, and falls back to using `/dev/urandom` (or `/dev/random` depending on the codepath) if the call to `getrandom` fails. This could be because the kernel does not support it, has not gathered enough entropy yet, etc.

Here's the interesting part: different host kernels will have different results here, and the result actually affects more than just `rustc` – Python, for example, also complained when it was unable to initialize randomness. I was able to run `rustc` using the version of `run-lang` from JRaspass/code-golf@b4aeff18b9fd7f8c56a5ab641d28f5f845357d05 on macOS, which uses the following kernel:
```
Linux f5f1bf6aa4a7 4.9.184-linuxkit #1 SMP Tue Jul 2 22:58:16 UTC 2019 x86_64 Linux
```
On the other hand, when I attempted to run on a CentOS 7-7 VM, both Python + `rustc` failed to initialize with `getrandom` and attempted to use `/dev/urandom` and/or `/dev/random`. This system has the kernel
```
Linux 8f5f10111ed1 3.10.0-514.el7.x86_64 #1 SMP Tue Nov 22 16:42:41 UTC 2016 x86_64 Linux
```
On this system, I was able to get everything working again by mounting a tmpfs to `/dev` and using `mknod` to create `urandom` and `random` as you suggested.

Based on all this, it seems as though the requirement to create `/dev` is host-dependent, and given you've been running Python + other languages without a problem I imagine your production system's kernel supports `getrandom` during the app initialization, which means all these `run-lang` changes are probably not necessary at all to run in production. However if you want increased portability it seems like creating the `/dev/{u}random` character block files is valuable, and in either case I am sure you will want to test out the changes before deploying them to production.

So, I will push the changes which create `/dev/{u}random` to this branch, but I have also created a branch ian-h-chamberlain/code-golf@073c4cbf86d5abb8cbcab8e56840dd14b1a8541d which reverts those changes, so you may test that if you'd prefer to keep `run-lang` more specific to your host system. Let me know and I can update the PR accordingly.

If you want a Rust program to test here is a simple FizzBuzz implementation:
```rust
fn main(){for i in 1..101{println!("{}",if i%15==0{"FizzBuzz"}else
if i%3==0{"Fizz"}else if i%5==0{"Buzz"}else{print!("{}",i);""})}}
```
If `rustc` fails to initialize entropy even the trivial program `fn main(){}` will fail to compile and will display a backtrace indicating something about `/dev/urandom` inaccessible.

Thanks for taking the time to look at all this – I guess in the end it wasn't so simple after all!

JRaspass commented:
Right let's give this a go, whee! Excellent work @ian-h-chamberlain :+1:

JRaspass commented:
Live :-) https://code-golf.io/scores/all-holes/rust

ian-h-chamberlain commented:
@JRaspass  Awesome, thanks! Already several scores on the board 😄
